### DISCUSSION TITLE
What a cool project! Here's what it needs to succeed...

### DISCUSSION THREAD

**Discussion by atrauzzi** [microsoft/aspire#1104] [Ideas]
I was thrilled to see the announcement of Aspire and seeing some of the initial demos.  This is a neat - ecosystem specific - take on some concepts I wish I had time to build. As such, I'm very grateful to the effort that's been put in so far 👏.  Like other .NET creations, I'm keen to see this succeed, although am also quick to spot gaps that Microsoft doesn't address that result in long term blind spots and lack of traction.

_This thread may get a lot of drive-by comments over the months as people wrap their heads around the concept, feel it out for themselves and begin to discover the same gaps._

Without further ado, here's what I think some immediate near-term priorities for Aspire should be:

 - Get the output json format ratified or recognized as some kind of standard
   - I'm not sure what body would be appropriate, but this could maybe even just be some kind of public support from other tooling makers
 - Get as many vendors as possible onboard to contribute local dev abstractions of their services and offerings -- and where the vendors aren't willing, Microsoft must be willing to implement it themselves out of good will and a desire to attract usage
   - For example, how about a [Stripe CLI Webhooks server](https://stripe.com/docs/stripe-cli)?
   - What about things like proprietary non-Azure or non-first-party, but defacto cloud offerings?  Google Cloud Tasks emulator comes to mind.
 - [Support multiple repositories](https://github.com/dotnet/aspire/issues/829) and make superdang sure it's extremely easy to manage
   - I think there's some cool innovation possible here where maybe you just check out the Aspire repo and then it piggybacks your local credentials to clone all the missing dependency repositories in a github-like folder structure, relative to itself.  This would be amazing for onboarding experiences!!
 - Offer some helpful tooling and libraries around the manifest for crazy places that still hand roll everything
   - This one I wish I didn't have to suggest, but I'm sure we all know of many places that think hand-rolling scripts that spit out kubernetes, helm and deployment templates based on some haphazard internal service descriptions is the way to go.  Aspire could possibly help these kinds of places round off and handle gaps.
 - Don't just be another tech that claims to be cross-cloud, but conspicuously only ships a handful of toy Azure targets
 - Test on podman, running on Linux, Windows and macOS. Not/not just docker desktop

This project is ambitious and I honestly hope it can grow to really be a useful tool for the position it's taken up. That said, in its current form today, Aspire makes for nice stage demos and edutainment, but isn't a tool I could justify adopting across the teams I oversee as a developer & architect.

If Microsoft takes this in the wrong directions too early, or ignores what I've listed above, I do see it sharing a fate similar to the likes of MAUI and I'll say it again: I'd love to see Aspire succeed!

**Comment by davidfowl** [comment]
> Get the output json format ratified or recognized as some kind of standard
I'm not sure what body would be appropriate, but this could maybe even just be some kind of public support from other tooling makers

I think we need to prove it out before we do that. It's premature at this point to even broach this topic until we have at least 3 or 4 instances of this pattern scaling to other systems.

> Get as many vendors as possible onboard to contribute local dev abstractions of their services and offerings -- and where the vendors aren't willing, Microsoft must be willing to implement it themselves out of good will and a desire to attract usage

Yes this is a goal. Though we want to post v1 have people be able to write their own resources and components without having to contribute to this repository.

> https://github.com/dotnet/aspire/issues/829 and make superdang sure it's extremely easy to manage

I'm looking forward to seeing what patterns evolve from the primitives we end up building to enable people to build this in previews. The story here will not be smooth for the first version but it will be possible.

> Offer some helpful tooling and libraries around the manifest for crazy places that still hand roll everything

I love this idea, but I don't know how to help in situations where these teams aren't using a system we can target. Do we push them to kubernetes? Do we tell them to use docker compose for deployment? It's unclear to me right now how we reach folks on prem that are not targeting consistent programmable infrastructure without taking on the burden of being another production orchestrator (which we have desire to do).

> Don't just be another tech that claims to be cross-cloud, but conspicuously only ships a handful of toy Azure targets

Asbolutely. I'm reallying hoping we can get some [basic AWS support soon](https://github.com/dotnet/aspire/pull/2225)!

> Test on podman, running on Linux, Windows and macOS. Not/not just docker desktop

https://github.com/dotnet/aspire/issues/827

> This project is ambitious and I honestly hope it can grow to really be a useful tool for the position it's taken up. That said, in its current form today, Aspire makes for nice stage demos and edutainment, but isn't a tool I could justify adopting across the teams I oversee as a developer & architect.

It is, we're trying to make it usable for real world scenarios. The team is learning quickly but there's still much to learn. We're laser focused on a couple of key vertical scenarios that are helping us flesh out the right primitives for v1.
